### ISSUE TITLE
Revise the design, implementation, and testing of the loop construct.

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#505]
The loop construct allows iteration through a computed set like so:

```
def f(x: Int, y: Int; z: Int): Set[Int] = List.toSet(x :: y :: z :: Nil)

rel R(x: Int).
R(x) :- x <- f(1, 2, 3)
```

but its current design is brittle and untested.

Depends on #502.

**Comment by magnus-madsen** [comment]
We should reintroduce support for loop by requiring the looping function to return an array.

**Comment by magnus-madsen** [comment]
No longer relevant.
